File: ts/project/.sdk/tm/scala/feature/LogFeature.scala
package SCALAPACKAGE.feature

import java.util.{Map => JMap}
import java.util.logging.{Level, Logger}
import SCALAPACKAGE.core.{Context, SdkClient}

// Hook logging via java.util.logging (JDK built-in; zero dependencies).
class LogFeature extends BaseFeature("log", "0.0.1", true) {

  private var client: SdkClient = null
  private var options: JMap[String, Object] = null
  private var logger: Logger = null

  override def init(ctx: Context, options: JMap[String, Object]): Unit = {
    this.client = ctx.client
    this.options = options

    options.get("active") match { case b: java.lang.Boolean => this.active = b.booleanValue(); case _ => }

    if (this.active) {
      options.get("logger") match {
        case l: Logger => this.logger = l
        case _ =>
          var level = Level.INFO
          options.get("level") match {
            case "debug" => level = Level.FINE
            case "warn" => level = Level.WARNING
            case "error" => level = Level.SEVERE
            case _ =>
          }
          this.logger = Logger.getLogger("ProjectNameSDK.log")
          this.logger.setLevel(level)
      }
    }
  }

  override def postConstruct(ctx: Context): Unit = loghook("PostConstruct", ctx, "")
  override def postConstructEntity(ctx: Context): Unit = loghook("PostConstructEntity", ctx, "")
  override def setData(ctx: Context): Unit = loghook("SetData", ctx, "")
  override def getData(ctx: Context): Unit = loghook("GetData", ctx, "")
  override def setMatch(ctx: Context): Unit = loghook("SetMatch", ctx, "")
  override def getMatch(ctx: Context): Unit = loghook("GetMatch", ctx, "")
  override def prePoint(ctx: Context): Unit = loghook("PrePoint", ctx, "")
  override def preSpec(ctx: Context): Unit = loghook("PreSpec", ctx, "")
  override def preRequest(ctx: Context): Unit = loghook("PreRequest", ctx, "")
  override def preResponse(ctx: Context): Unit = loghook("PreResponse", ctx, "")
  override def preResult(ctx: Context): Unit = loghook("PreResult", ctx, "")

  private def loghook(hook: String, ctx: Context, level0: String): Unit = {
    if (this.logger == null) return
    val level = if ("" == level0) "info" else level0

    val msg = new StringBuilder("hook=" + hook)
    if (ctx.op != null) msg.append(" op=").append(ctx.op.name)
    if (ctx.spec != null) msg.append(" spec=").append(ctx.spec.method).append(" ").append(ctx.spec.path)

    level match {
      case "debug" => this.logger.fine(msg.toString)
      case "warn" => this.logger.warning(msg.toString)
      case "error" => this.logger.severe(msg.toString)
      case _ => this.logger.info(msg.toString)
    }
  }
}

File: ts/project/.sdk/tm/java/feature/LogFeature.java
package JAVAPACKAGE.feature;

import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

import JAVAPACKAGE.core.Context;
import JAVAPACKAGE.core.SdkClient;

/** Hook logging via java.util.logging (JDK built-in; zero dependencies). */
public class LogFeature extends BaseFeature {

  private SdkClient client;
  private Map<String, Object> options;
  private Logger logger;

  public LogFeature() {
    super("log", "0.0.1", true);
  }

  @Override
  public void init(Context ctx, Map<String, Object> options) {
    this.client = ctx.client;
    this.options = options;

    Object active = options.get("active");
    if (active instanceof Boolean) {
      this.active = (Boolean) active;
    }

    if (this.active) {
      if (options.get("logger") instanceof Logger) {
        this.logger = (Logger) options.get("logger");
      }
      else {
        Level level = Level.INFO;
        Object lvl = options.get("level");
        if ("debug".equals(lvl)) {
          level = Level.FINE;
        }
        else if ("warn".equals(lvl)) {
          level = Level.WARNING;
        }
        else if ("error".equals(lvl)) {
          level = Level.SEVERE;
        }

        this.logger = Logger.getLogger("ProjectNameSDK.log");
        this.logger.setLevel(level);
      }
    }
  }

  @Override
  public void postConstruct(Context ctx) {
    loghook("PostConstruct", ctx, "");
  }

  @Override
  public void postConstructEntity(Context ctx) {
    loghook("PostConstructEntity", ctx, "");
  }

  @Override
  public void setData(Context ctx) {
    loghook("SetData", ctx, "");
  }

  @Override
  public void getData(Context ctx) {
    loghook("GetData", ctx, "");
  }

  @Override
  public void setMatch(Context ctx) {
    loghook("SetMatch", ctx, "");
  }

  @Override
  public void getMatch(Context ctx) {
    loghook("GetMatch", ctx, "");
  }

  @Override
  public void prePoint(Context ctx) {
    loghook("PrePoint", ctx, "");
  }

  @Override
  public void preSpec(Context ctx) {
    loghook("PreSpec", ctx, "");
  }

  @Override
  public void preRequest(Context ctx) {
    loghook("PreRequest", ctx, "");
  }

  @Override
  public void preResponse(Context ctx) {
    loghook("PreResponse", ctx, "");
  }

  @Override
  public void preResult(Context ctx) {
    loghook("PreResult", ctx, "");
  }

  private void loghook(String hook, Context ctx, String level) {
    if (this.logger == null) {
      return;
    }

    if ("".equals(level)) {
      level = "info";
    }

    StringBuilder msg = new StringBuilder("hook=" + hook);

    if (ctx.op != null) {
      msg.append(" op=").append(ctx.op.name);
    }

    if (ctx.spec != null) {
      msg.append(" spec=").append(ctx.spec.method).append(" ").append(ctx.spec.path);
    }

    switch (level) {
      case "debug":
        this.logger.fine(msg.toString());
        break;
      case "warn":
        this.logger.warning(msg.toString());
        break;
      case "error":
        this.logger.severe(msg.toString());
        break;
      default:
        this.logger.info(msg.toString());
        break;
    }
  }
}
